_delphi_utils_python/delphi_utils/geomap.py

```diff
@@ -595,3 +595,59 @@ def get_geo_values(self, geo_type):
             crosswalk = pd.read_csv(stream, dtype=str)
             self.geo_lists[geo_type] = set(crosswalk[geo_type])
             return self.geo_lists[geo_type]
+
+    def get_geos_within(self, container_geocode, child_geocode_type,container_geocode_type):
+        """
+        Return all contained/child regions.
+
+        Given a container geocode (e.g. "ca" for California, a state)
+        and an enclosed child geocode type (e.g. "county").
+        return a set of container geocode value of the specified type that
+        lie within the specified geo code (e.g. all counties within California)
+        Support these 3 types:
+            - all states within a nation
+            - all counties within a state
+            - all states within an hhs region
+
+        Parameters
+        ----------
+        container_geocode: str of identity of nation/state/hhs
+            "fips" for return container_geocode of state
+            "state_id" for return container_geocode of nation and hhs
+        child_geocode_type: str
+            One of "state","county"
+        container_geocode_type: str
+            One of "state","nation","hhs"
+
+        Returns
+        -------
+        Set of geo ids of the specified type that lie within the specified container geocode,
+        all in string format.
+        """
+        if child_geocode_type not in ("county","state"):
+            raise ValueError("child_geocode_type must be one of state, nation and hhs")
+        get_geo_values=set()
+        if child_geocode_type=="state":
+            if container_geocode_type=="nation" and container_geocode=="us":
+                crosswalk = self._load_crosswalk_from_file("state", "state")
+                get_geo_values=set(crosswalk["state_id"])
+            if container_geocode_type=="hhs":
+                crosswalk_fips_to_hhs = self._load_crosswalk_from_file("fips", "hhs")
+                crosswalk_fips_to_state = self._load_crosswalk_from_file("fips", "state")
+                fips_hhs=[]
+                for i in range(1,len(crosswalk_fips_to_hhs["hhs"])):
+                    if crosswalk_fips_to_hhs["hhs"][i]==container_geocode:
+                        fips_hhs.append(crosswalk_fips_to_hhs["fips"][i])
```

Comment:
This for-loop gets the job done and I think it's correct! However, I want to show you a really cool thing in Pandas that lets you do this kind of selection in a really slick way (I think you're gonna like how clean this is):
`crosswalk_fips_to_hhs[crosswalk_fips_to_hss["hhs"] == container_geocode]`

I'll walk through what's happening here:
1. Since `crosswalk_fips_to_hhs` is a `pd.DataFrame`, the result of `crosswalk_fips_to_hss["hhs"] == container_geocode` is a Pandas boolean Series with the same index as `crosswalk_fips_to_hhs` and True/False values if the original row had an "hhs" column equal to `container_geocode`.
2. Now the really cool thing is that you can feed `boolean_series = (crosswalk_fips_to_hss["hhs"] == container_geocode)` back into `crosswalk_fips_to_hss` and just get exactly the rows that were True.

[Check out Panda's more general guide](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#boolean-indexing) for more examples on these sorts of things.

Try refactoring these lines using these methods!

PS. Also once you get a dataframe of all the fips codes that have a particular hhs, you will need to use [`pd.DataFrame.isin`](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-with-isin) to select states with fips codes in that dataframe.

Reply:
That's really cool. Thanks for the explanation!